File: day-4/go/part2.go
package day4

import (
	"fmt"
	"os"
	"strings"
	"time"
)

// filter out a string from an array of strings
func filter_with(arr *[]string, str string) {
	for i := 0; i < len(*arr)-1; i++ {
		if (*arr)[i] == str {
			*arr = append((*arr)[:i], (*arr)[i+1:]...)
		}
	}
}

func (d Day4) Part2(filename *string) string {
	content, _ := os.ReadFile(*filename)
	lines := strings.Split(string(content), "\n")

	// clock the time
	start := time.Now()

	var cards map[int]int = map[int]int{}

	for card_num, line := range lines {
		var all_numbers = strings.Split(line, ": ")
		var split_numbers = strings.Split(all_numbers[1], " | ")
		var winning_numbers = strings.Split(strings.Trim(split_numbers[0], " "), " ")
		var own_numbers = strings.Split(strings.Trim(split_numbers[1], " "), " ")

		// filter out the spaces for the numbers
		filter_with(&winning_numbers, "")
		filter_with(&own_numbers, "")

		var count_winners = 0

		// create a map to hold the winning numbers
		winners_num_map := map[string]bool{}
		for _, winning_number := range winning_numbers {
			winners_num_map[winning_number] = true
		}

		// loop through the own numbers and see if they match
		for _, own_number := range own_numbers {
			if winners_num_map[own_number] {
				count_winners++
			}
		}

		// check if we won!
		if count_winners > 0 {
			for i := 1; i <= count_winners; i++ {
				cards[card_num+i] += 1 + cards[card_num]
			}
		}

		// add the card to self
		cards[card_num] += 1
	}

	// count the total number of scratch cards
	var total_scratch_cards = 0

	for _, scratch_count := range cards {
		total_scratch_cards += scratch_count
	}

	// print the time it took to run
	fmt.Println(time.Since(start))

	return fmt.Sprint(total_scratch_cards)
}

File: day-04/cpp/part2.cpp
// To run do: c++ part2.cpp -Wall -std=c++17 -o output2.out && ./output2.out
#include <string>
#include <vector>
#include <fstream>
#include <iostream>
#include <map>
#include "../../utils/utils.cpp"

using namespace std;

// filter empty strings from vector
void filter_empty_strings(std::vector<std::string> *str_vector)
{
    for (int i = 0; i < str_vector->size(); i++)
    {
        if ((*str_vector)[i] == "")
        {
            str_vector->erase(str_vector->begin() + i);
        }
    }
}

int main()
{
    // Initialize a vector to store the file input and set the input file name.
    std::string fileName = "../input.txt";
    std::vector<std::string> lines = read_file(fileName);

    // initialize map
    map<int, int> scratch_map;

    for (int scratch_num = 0; scratch_num < lines.size(); scratch_num++)
    {
        // parse out the winning and own numbers
        std::vector<std::string> words = splitter(lines[scratch_num], ": ");
        std::vector<std::string> split_numbers = splitter(words[1], " | ");
        std::vector<std::string> winning_numbers = splitter(trim(split_numbers[0]), " ");
        std::vector<std::string> own_numbers = splitter(trim(split_numbers[1]), " ");

        // filter out empty strings
        filter_empty_strings(&winning_numbers);
        filter_empty_strings(&own_numbers);

        // count the total number of winners
        int count_winners = 0;
        for (int j = 0; j < winning_numbers.size(); j++)
        {
            for (int k = 0; k < own_numbers.size(); k++)
            {
                if (winning_numbers[j] == own_numbers[k])
                {
                    count_winners++;
                }
            }
        }

        // check if we won
        if (count_winners > 0)
        {
            // if we won, add the count to the next count_winners scratch_maps
            for (int j = 1; j <= count_winners; j++)
            {
                scratch_map[scratch_num + j] += 1 + scratch_map[scratch_num];
            }
        }

        // add the scratch to self
        scratch_map[scratch_num] += 1;
    }

    // loop through map and add up all the values
    int scratch_total = 0;
    for (auto scratch : scratch_map)
    {
        scratch_total += scratch.second;
    }

    // Print the result.
    std::cout << "Result: " << scratch_total << std::endl;
}
